File: 02_LearnDSP-MATLAB/Sec04_PracticalMethods/case03_Smooth.m
% Smooth data

% XiaoCY 2021-02-15

%%
clear;clc

fs = 100;
t = (0:1/fs:20)';
x0 = cos(2*pi*t);
xn = randn(size(t))*0.2;
x = x0+xn;

%% five-spot triple smoothing method
m = 21;                         % smooth times

a = x;
y1 = x;
N = length(x);
for k = 1:m
    y1(1) = (69*a(1) + 4*(a(2)+a(4)) - 6*a(3) - a(5))/70;
    y1(2) = (2*(a(1)+a(5)) + 27*a(2) + 12*a(3) - 8*a(4))/35;
    for j = 3:N-2
        y1(j) = (-3*(a(j-2)+a(j+2)) + 12*(a(j-1)+a(j+1)) + 17*a(j))/35;
    end
    y1(N-1) = (2*(a(N-4)+a(N)) - 8*a(N-3) + 12*a(N-2) + 27*a(N-1))/35;
    y1(N) = (-a(N-4) + 4*(a(N-3)+a(N-1)) - 6*a(N-2) + 69*a(N))/70;
    a = y1;
end

%% built-in function
y21 = smooth(x,21,'moving');
y22 = smooth(x,21,'lowess');
y23 = smooth(x,21,'loess');
y24 = smooth(x,21,'sgolay');

y3 = sgolayfilt(x,5,21);

%%
figure
plot(t,x,'DisplayName','origin')
hold on
grid on
plot(t,y1,'DisplayName',sprintf('mean53(%d)',m))
plot(t,y21,'DisplayName','Smooth-moving')
plot(t,y22,'DisplayName','Smooth-lowess')
plot(t,y23,'DisplayName','Smooth-loess')
plot(t,y24,'DisplayName','Smooth-sgolay')
plot(t,y3,'DisplayName','Savitzky-Golay')
legend
xlabel('Time (s)')
ylabel('Signal')

figure
iLPSD([x,y1,y21,y22,y23,y24,y3],fs)
xlabel('Frequency (Hz)')
ylabel('PSD')
legend('origin',sprintf('mean53(%d)',m),'Smooth-moving',...
    'Smooth-lowess','Smooth-loess','Smooth-sgolay','Savitzky-Golay')

File: 03_LearnDSP-Python/Sec04_PracticalMethods/case03_Smooth.py
'''
Smooth data

XiaoCY 2021-02-15
'''

#%%
import numpy as np
import matplotlib.pyplot as plt
import scipy.signal as signal
import springlib as sp

fs = 100.
t = np.arange(0,20,1/fs)
x0 = np.cos(2*np.pi*t)
xn = np.random.randn(len(t))*0.2
x = x0+xn

#%% five-spot triple smoothing method
m = 21                          # smooth times

a = x.copy()
y1 = x.copy()
N = len(x)
for k in range(m):
    y1[0] = (69*a[0] + 4*(a[1]+a[3]) - 6*a[2] - a[4])/70
    y1[1] = (2*(a[0]+a[4]) + 27*a[1] + 12*a[2] - 8*a[3])/35
    for j in range(2,N-2):
        y1[j] = (-3*(a[j-2]+a[j+2]) + 12*(a[j-1]+a[j+1]) + 17*a[j])/35
    y1[N-2] = (2*(a[N-5]+a[N-1]) - 8*a[N-4] + 12*a[N-3] + 27*a[N-2])/35
    y1[N-1] = (-a[N-5] + 4*(a[N-4]+a[N-2]) - 6*a[N-3] + 69*a[N-1])/70
    a = y1.copy()

#%% built-in function
# No `smooth` function in Python
y3 = signal.savgol_filter(x,21,5)

#%%
plt.figure()
plt.plot(t,x)
plt.plot(t,y1)
plt.plot(t,y3)
plt.grid()
plt.legend(('origin','mean53('+str(m)+')','Savitzky-Golay'))
plt.xlabel('Time (s)')
plt.ylabel('Signal')
plt.show()

px,f = sp.lpsd(x,fs)
py1,_ = sp.lpsd(y1,fs)
py3,_ = sp.lpsd(y3,fs)
plt.figure()
plt.loglog(f,np.sqrt(px))
plt.loglog(f,np.sqrt(py1))
plt.loglog(f,np.sqrt(py3))
plt.grid()
plt.legend(('origin','mean53('+str(m)+')','Savitzky-Golay'))
plt.xlabel('Frequency (Hz)')
plt.ylabel('PSD')
plt.show()